### PR TITLE
Simplify construction and updates to DiagnosticAnalysisResult

### PR DESCRIPTION
**Pull request by sharwell**
:memo: Extracted from #30733

### CODE REVIEW

File: src/Workspaces/Core/Portable/Diagnostics/DiagnosticAnalysisResult.cs [hunk 1]
```diff
@@ -82,11 +71,81 @@ internal struct DiagnosticAnalysisResult
...
+        }
+
+        public static DiagnosticAnalysisResult CreateFromBuild(Project project, ImmutableArray<DiagnosticData> diagnostics)
+        {
+            // we can't distinguish locals and non locals from build diagnostics nor determine right snapshot version for the build.
+            // so we put everything in as semantic local with default version. this lets us to replace those to live diagnostics when needed easily.
+            var version = VersionStamp.Default;
+
+            // filter out any document that doesn't support diagnostics
+            var group = diagnostics.GroupBy(d => d.DocumentId);
```
Comment (heejaechang) [inline src/Workspaces/Core/Portable/Diagnostics/DiagnosticAnalysisResult.cs]: so, we will put the diagnostic support check later?

Comment (sharwell) [reply src/Workspaces/Core/Portable/Diagnostics/DiagnosticAnalysisResult.cs]: ➡️ That is correct. The current pull request is just a refactoring. The `SupportsDiagnostics` method doesn't exist in the target branch yet, and this location will be minimally updated to use it when it's added.

File: src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.Executor.cs [hunk 2]
```diff
@@ -165,15 +165,7 @@ public IEnumerable<DiagnosticData> ConvertToLocalDiagnostics(Document targetDocu
...
-                var newCompilerAnalysisResult = new DiagnosticAnalysisResult(
-                    analysisResult.ProjectId,
-                    analysisResult.Version,
-                    analysisResult.SyntaxLocals,
-                    semanticLocals: ImmutableDictionary<DocumentId, ImmutableArray<DiagnosticData>>.Empty,
-                    nonLocals: ImmutableDictionary<DocumentId, ImmutableArray<DiagnosticData>>.Empty,
-                    others: ImmutableArray<DiagnosticData>.Empty,
-                    documentIds: null,
-                    fromBuild: false);
+                var newCompilerAnalysisResult = analysisResult.DropExceptSyntax();
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.Executor.cs]: Cute.   I'm also feelign taht a .WithNothingButSyntax might be a better name.  The reason for this (to me at least) is that it follows the 'With' namign pattern that clearly indicates that you need to use the result.  'DropXXX' sounds very imperative and mutable, and makes me think that actual instance is mutating itself.  

Thoughts?

File: src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.ProjectState.cs [hunk 3]
```diff
@@ -232,7 +232,7 @@ public async Task SaveAsync(Project project, DiagnosticAnalysisResult result)
             public void ResetVersion()
             {
                 // reset version of cached data so that we can recalculate new data (ex, OnDocumentReset)
-                _lastResult = new DiagnosticAnalysisResult(_lastResult.ProjectId, VersionStamp.Default, _lastResult.DocumentIds, _lastResult.IsEmpty, _lastResult.FromBuild);
+                _lastResult = _lastResult.Reset();
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.ProjectState.cs]: Same concern aobut this name.  But i don't have anything better just yet.

File: src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.Executor.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.Executor.cs]: If you do like 'Drop', i would be ok with WithAllDroppedExceptSyntax.

Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.Executor.cs]: Alternatively.  could this be `analysisResult.Reset().WithSyntax(...)`?

File: src/Workspaces/Core/Portable/Diagnostics/DiagnosticAnalysisResult.cs [hunk 4]
```diff
@@ -82,11 +71,81 @@ internal struct DiagnosticAnalysisResult
...
+        }
+
+        public static DiagnosticAnalysisResult CreateFromSerialization(
+            ProjectId projectId,
+            VersionStamp version,
+            ImmutableDictionary<DocumentId, ImmutableArray<DiagnosticData>> syntaxLocalMap,
+            ImmutableDictionary<DocumentId, ImmutableArray<DiagnosticData>> semanticLocalMap,
+            ImmutableDictionary<DocumentId, ImmutableArray<DiagnosticData>> nonLocalMap,
+            ImmutableArray<DiagnosticData> others,
+            ImmutableHashSet<DocumentId> documentIds = null)
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Diagnostics/DiagnosticAnalysisResult.cs]: this being null seems super weird.  why would serialization not have this info?

File: src/Workspaces/Core/Portable/Diagnostics/DiagnosticAnalysisResult.cs [hunk 5]
```diff
@@ -82,11 +71,81 @@ internal struct DiagnosticAnalysisResult
...
+        public static DiagnosticAnalysisResult CreateFromBuilder(DiagnosticAnalysisResultBuilder builder)
         {
+            return CreateFromSerialization(
+                builder.Project.Id,
+                builder.Version,
+                builder.SyntaxLocals,
+                builder.SemanticLocals,
+                builder.NonLocals,
+                builder.Others,
+                builder.DocumentIds);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Diagnostics/DiagnosticAnalysisResult.cs]: seems weird as i would just expect the builder to have a .ToResult method on it.

File: src/Workspaces/Core/Portable/Diagnostics/DiagnosticAnalysisResult.cs [hunk 6]
```diff
@@ -121,6 +180,29 @@ public DiagnosticAnalysisResult ToAggregatedForm()
             return new DiagnosticAnalysisResult(ProjectId, Version, DocumentIds, IsEmpty, FromBuild);
         }
 
+        public DiagnosticAnalysisResult UpdateAggregatedResult(VersionStamp version, DocumentId documentId, bool fromBuild)
+        {
+            return new DiagnosticAnalysisResult(ProjectId, version, DocumentIdsOrEmpty.Add(documentId), isEmpty: false, fromBuild: fromBuild);
+        }
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Diagnostics/DiagnosticAnalysisResult.cs]: I'm not sure i understand the Aggregated bit here.

File: src/Workspaces/Core/Portable/Diagnostics/DiagnosticAnalysisResult.cs [hunk 7]
```diff
@@ -121,6 +180,29 @@ public DiagnosticAnalysisResult ToAggregatedForm()
...
+
+        public DiagnosticAnalysisResult Reset()
+        {
+            return new DiagnosticAnalysisResult(ProjectId, VersionStamp.Default, DocumentIds, IsEmpty, FromBuild);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Diagnostics/DiagnosticAnalysisResult.cs]: this is subtle to me.  the fact that 'FromBuild' and 'IsEmpty' are preserved are not necessairly things i would expect.

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Diagnostics/DiagnosticAnalysisResult.cs]: also, it's not clear to me if 'IsEmpty' is actually correct here.  IsEmpty above is computed as ` IsEmpty = DocumentIds.IsEmpty && _others.IsEmpty;`.  But we don't pass '_others' here.  So we may have been non-empty before, but we should be empty now.  Soemthing seems smelly here.

File: src/Workspaces/Core/Portable/Diagnostics/DiagnosticAnalysisResult.cs [hunk 8]
```diff
@@ -121,6 +180,29 @@ public DiagnosticAnalysisResult ToAggregatedForm()
...
+        {
+            return new DiagnosticAnalysisResult(
+               ProjectId,
+               Version,
+               SyntaxLocals,
+               semanticLocals: ImmutableDictionary<DocumentId, ImmutableArray<DiagnosticData>>.Empty,
+               nonLocals: ImmutableDictionary<DocumentId, ImmutableArray<DiagnosticData>>.Empty,
+               others: ImmutableArray<DiagnosticData>.Empty,
+               documentIds: null,
+               fromBuild: false);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Diagnostics/DiagnosticAnalysisResult.cs]: can we have comments about the 'fromBuild' being switched to 'false'?  it's unclear to me when that should be changing, and it's a subtlety i think a caller coudl get wrong.

File: src/Workspaces/Core/Portable/Diagnostics/DiagnosticAnalysisResult.cs [hunk 9]
```diff
@@ -82,11 +71,81 @@ internal struct DiagnosticAnalysisResult
             IsEmpty = DocumentIds.IsEmpty && _others.IsEmpty;
         }
 
-        public DiagnosticAnalysisResult(DiagnosticAnalysisResultBuilder builder) :
-            this(builder.Project.Id, builder.Version,
-                builder.SyntaxLocals, builder.SemanticLocals, builder.NonLocals, builder.Others,
-                builder.DocumentIds, fromBuild: false)
+        public static DiagnosticAnalysisResult CreateEmpty(ProjectId projectId, VersionStamp version)
+        {
+            return new DiagnosticAnalysisResult(
```
Comment (genlu) [inline src/Workspaces/Core/Portable/Diagnostics/DiagnosticAnalysisResult.cs]: i find it a bit confusing that everything here is empty but `isEmpty == false`, whereas the one created by `CreateInitialResult` have all default values but `isEmpty == true`. Maybe consider changing it to `IsDefault` or something like that?

**Comment by sharwell** [comment]
retest this please

**Comment by sharwell** [comment]
Verified this pull request is correct with the changes through bf32f41ab80ce173521a81eb7d35ecfba99e62c8

**Pull request merged**